src/lib/SaferNames/Name.hs

```diff
@@ -4,288 +4,206 @@
-                                        Right _    -> Nothing
+(<>>) :: Env v i o -> EnvFrag v i i' o -> Env v i' o
+(<>>) (Env f frag) (EnvFrag frag') = Env f $ concatNameMaps frag frag'
+
+infixl 1 <.>
+(<.>) :: EnvFrag v i1 i2 o -> EnvFrag v i2 i3 o -> EnvFrag v i1 i3 o
+(<.>) (EnvFrag env) (EnvFrag env') = EnvFrag (concatNameMaps env env')
+
+emptyEnvFrag :: EnvFrag v i i o
+emptyEnvFrag = EnvFrag emptyNameMap
+
+lookupEnvFrag :: EnvFrag v i i' o -> Name s i' -> Either (Name s i) (v s o)
+lookupEnvFrag (EnvFrag m) name =
+  case projectName (nameMapNames m) name of
+    Left  name' -> Left name'
+    Right name' -> Right $ lookupNameMap m name'
+
+lookupScope :: Scope n -> Name s n -> s n
+lookupScope (Scope m) v = fromIdE $ lookupNameMap m v
+
+appendScope :: Scope n -> ScopeFrag n l -> Scope l
+appendScope (Scope m1) (ScopeFrag m2) =
+  Scope $ extendNameMap (injectNames (nameMapNames m2) m1) m2
 
 -- === common scoping patterns ===
 
 data Abs (binder::B) (body::E) (n::S) where
   Abs :: binder n l -> body l -> Abs binder body n
 
-data FreshAbs (binder::B) (body::E) (n::S) where
-  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
-
 data NestPair (b1::B) (b2::B) (n::S) (l::S) where
   NestPair :: b1 n l -> b2 l l' -> NestPair b1 b2 n l'
 
 data Nest (binder::B) (n::S) (l::S) where
   Nest  :: binder n h -> Nest binder h l -> Nest binder n l
   Empty ::                                  Nest binder n n
 
-type PlainBinderList = Nest PlainBinder
-
-infixl 7 :>
-data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
-  (:>) (b n l) (ann n)
-  deriving (Show)
+type EmptyAbs b = Abs b UnitE :: E
+pattern EmptyAbs :: b n l -> EmptyAbs b n
+pattern EmptyAbs bs = Abs bs UnitE
+
+-- === Injections and projections ===
+
+newtype Ext (n::S) (l::S) = ExtVal { fromExtVal :: NameSet (n:=>:l) }
+
+instance Category Ext where
+  id = ExtVal emptyNameSetFrag
+  ext1 . ext2 = ExtVal $ concatNameSets (fromExtVal ext2) (fromExtVal ext1)
+
+class BindsNames (b :: B) where
+  toExtVal :: b n l -> Ext n l
+
+instance BindsNames Ext where
+  toExtVal = id
+
+toNameSet :: BindsNames b => b n l -> NameSet (n:=>:l)
+toNameSet b = fromExtVal $ toExtVal b
+
+inject :: BindsNames b => InjectableE e => Distinct l => b n l -> e n -> e l
+inject ext x = injectNames (toNameSet ext) x
+
+-- like inject, but uses the SopeReader monad for its `Distinct` proof
+injectM :: ScopeReader m => BindsNames b => InjectableE e => b n l -> e n -> m l (e l)
+injectM b e = withDistinct $ return $ inject b e
+
+fromConstAbs :: ScopeReader m => MonadErr1 m => BindsNames b => HasNamesE e
+             => Abs b e n -> m n (e n)
+fromConstAbs (Abs b e) = do
+  let t = newNameTraversal \name ->
+             case projectName (toNameSet b) name of
+               Left name' -> Just name'
+               Right _    -> Nothing
+  scope <- askScope
+  case traverseNamesE scope t e of
+    Nothing -> throw EscapedNameErr ""
```

Comment:
Add more context? Otherwise we'll have a great time debugging those. Also, it seems a little inefficient that we traverse the expression every time we want to use this, because I expect that we'll be doing this all over the place. If we don't have a better idea then let's at least mark this as a potential hot spot.

Reply:
I can't think of an alternative to traversing the expression. We definitely need to be conscious of its cost when we use it. I'll make a note.

Regarding context, I was planning to at least add the leaked names themselves to the context. Other than that, I think it's best to add context from higher up in the call stack, where we have more information about what's going on. A function like this is so general that it's not clear what more context we could add.

Reply:
Actually, adding the name to the context was an easy switch from `Maybe` to `Either Err`. Hooray for monads!